### PR TITLE
[Dashboard] Only auto find port in ray.init() case

### LINKED ISSUE #14257: Ensure dashboard starts on the desired port
### What is the problem?
We need a guarantee that the ray dashboard will get started on port 8265 when `ray stop` and `ray start` are called programmatically in quick succession.

*Ray version and other system information (Python version, TensorFlow version, OS):*

In Ray 1.2.0 (Python 3.7) using `ray stop --force`, fixes the problem the majority of the time but we're still seeing it get started on the undesired port about 1/2000 tries.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

Courtesy of @ijrsvt 
```
import subprocess
iteration = 0
while True:
    output = subprocess.run("ray stop --force; ray start --head > /dev/null", shell=True, capture_output=True)
    assert "8265" in output.stderr.decode()
    print("Started on 8265", " -- Iteration -- ", iteration, ")
    iteration += 1
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by simon-mo**
Closes #14257. This PR requires a API review (@pcmoritz) before merge.

Old behavior:
- `ray.init()` and `ray start` defaults to port 8625 and if that port is bound, Ray will find the next available port.
- Ray will skip the auto port finding only when the dashboard port is explicitly passed in.
- `ray stop; ray start --head` sometimes start dashboard in new port even though the old port should be freed. This is because the OS might hold on to that port longer.

New behavior:
- `ray.init()` still do the automatic free port finding, in accordance to the randomized redis port behavior
- `ray start` defaults to 8625, if the port if bound, `ray start` should fail.
- Ray will skip auto port finding only when the dashboard port is explicitly passed in. (unchanged)
- `ray stop; ray start --head` will always bind to 8625. If the port was still occupied by previous dashboard (held by the OS case), new dashboard will go ahead and serve that port 8625 by using SO_REUSEPORT and keep retrying to bind to that port.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
🤦‍♂️ there seems to be another PR that does similar things (dashboard automatically find a port by itself and talk to redis?), just merged yesterday. 
https://github.com/ray-project/ray/pull/13763/files

**Comment by fyrestone** [comment]
The PR #13763 moves the port selection logic to dashboard, and make sure the dashboard listens on a valid port if `ray.init()` success. The port selection logic is not changed.

I think it's better to set the default dashboard port to None and auto select port only if the dashboard port is None.

**Comment by fyrestone** [comment]
I can't find the changes for `scripts.py`. So, is the`ray start --head` without `--dashboard-port` fails if the port `8265` is occupied?

**Comment by simon-mo** [comment]
Yes. @fyrestone this the an API change, `ray start --head` will default to _only_ 8625. This makes the `ray start --head` behavior consistent (same redis port, same dashboard port).

**Review by fyrestone — APPROVED** [review]
LGTM

**Review by edoakes — APPROVED** [review]

File: python/ray/worker.py [hunk 1]
```diff
@@ -572,8 +572,9 @@ def init(
             localhost (127.0.0.1) or 0.0.0.0 (available from all interfaces).
             By default, this is set to localhost to prevent access from
             external machines.
-        dashboard_port: The port to bind the dashboard server to. Defaults to
-            8265.
+        dashboard_port(int, None): The port to bind the dashboard server to.
+            Defaults to 8265 and Ray will automatically find a free port if
+            8265 is not available.
```
Comment (edoakes) [inline python/ray/worker.py:577]:
```suggestion
            If not specified, Ray will automatically select a port starting at 8265 and incrementing until a free port is found.
```

**Review by edoakes — APPROVED** [review]
@simon-mo LGTM, but there seem to be a lot of tests failing.

**Comment by ijrsvt** [comment]
> new dashboard will go ahead and serve that port 8625 by using SO_REUSEPORT and keep retrying to bind to that port.

^ Where is this modified? I think this requires a change inside the dashboard code.

**Comment by fyrestone** [comment]
> > new dashboard will go ahead and serve that port 8625 by using SO_REUSEPORT and keep retrying to bind to that port.
> 
> ^ Where is this modified? I think this requires a change inside the dashboard code.
@simon-mo I think the description of this PR needs to be updated.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**